### DISCUSSION TITLE
Ignore Cross Section feature classes in Build Metadata

### DISCUSSION THREAD

**Discussion by vlatane-va-energy** [usgs/gems-tools-pro#98] [Ideas]
The max_bounding function in the GeMS_FGDCMetadata.py script checks "MapUnitPolys" and "ContactsAndFaults" layers to generate a bounding box.

  ```
def max_bounding(db_path):
      north = []
      south = []
      west = []
      east = []
      for layer in ogr.Open(db_path):
          # if layer.GetGeomType() != 100:
          if "MapUnitPolys" in layer.GetName() or "ContactsAndFaults" in layer.GetName():
              bounding = su.get_bounding(str(db_path), layer.GetName())
              north.append(bounding.find("northbc").text)
              south.append(bounding.find("southbc").text)
              west.append(bounding.find("westbc").text)
              east.append(bounding.find("eastbc").text)
  
      bounding.find("northbc").text = max(north)
      bounding.find("southbc").text = min(south)
      bounding.find("westbc").text = max(west)
      bounding.find("eastbc").text = min(east)
  
      return bounding
``` 
Cross Section MapUnitPolys and ContactsAndFaults feature classes also get caught by this function to identify the fcs used for the bounding box. Why does that matter? It's not uncommon to use "Unknown Coordinate System" on these CS feature classes, because they are often placed cartographically below the map instead of in 'true space'. Feature classes without a coordinate system will (correctly) trigger errors in the spatial utilities used by the Metadata Wizard to draw the bounding box and prevent the box from being drawn. The main Build Metadata tool ends if you cannot draw the box, leaving you without an xml entirely!

This could be avoided by ignoring the cross sections entirely in the max_bounding function. Because these aren't typically located in true space, they shouldn't be used to calculate the bounds of the data. We haven't worked out the exact language to pull out only the relevant MUP and CAF layers. Perhaps the best way is to set a new parameter and have users indicate which fcs should be considered for the bounding box.

**Comment by vlatane-va-energy** [comment]
Also, "e" in the code below acted as a red herring in our troubleshooting efforts. Maybe add something to indicate that this _might_ be the problem, but isn't always.

```
# find the bounding box that covers the extent of all features
try:
    if base_md.find("idinfo/spdom/bounding") is None:
        arcpy.AddMessage("  spdom")
        spdom = etree.Element("spdom")
        bounding = max_bounding(str(db_path))
        spdom.append(bounding)
        base_md.find("idinfo").append(spdom)
except Exception as error:
    e = """Could not calculate a bounding box.
    Set environment variable PROJ_LIB to location of proj.db and try again.
    See the ArcGIS Pro wiki for more information - https://github.com/usgs/gems-tools-pro/wiki/GeMS-Tools-Documentation#BuildMetadata"""
    arcpy.AddError(e)
    arcpy.AddError(error)
    sys.exit()
```

**Reply by ethoms-usgs** [reply]
You are absolutely right. Cross section features should be ignored for the bounding box calculation. I will work on a fix.

**Reply by ethoms-usgs** [reply]
Could you try this edit and tell me if it works? 
Add `if layer.GetSpatialRef():` (below) to def max_bounding()

```
def max_bounding(db_path):
    north = []
    south = []
    west = []
    east = []
    for layer in ogr.Open(db_path):
        # if layer.GetGeomType() != 100:
        if "MapUnitPolys" in layer.GetName() or "ContactsAndFaults" in layer.GetName():
            if layer.GetSpatialRef():
                bounding = su.get_bounding(str(db_path), layer.GetName())
                north.append(bounding.find("northbc").text)
                south.append(bounding.find("southbc").text)
                west.append(bounding.find("westbc").text)
                east.append(bounding.find("eastbc").text)

    bounding.find("northbc").text = max(north)
    bounding.find("southbc").text = min(south)
    bounding.find("westbc").text = max(west)
    bounding.find("eastbc").text = min(east)

    return bounding
```

Layers that have an 'Unknown' spatial reference return None for GetSpatialRef(), so the new if statement should filter those out.

**Reply by vlatane-va-energy** [reply]
Yes, that seems to have worked. I am getting a bounding box now on the same database that I was testing earlier, which has the Unknown Coordinate System layers. Thanks!

**Reply by ethoms-usgs** [reply]
Great! It's in v2.12.13.
